acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/OptimisticSyncPostMergeAcceptanceTest.java

```diff
@@ -0,0 +1,103 @@
+ */
+
+package tech.pegasys.teku.test.acceptance;
+
+import com.google.common.io.Resources;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import tech.pegasys.teku.infrastructure.time.SystemTimeProvider;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.test.acceptance.dsl.AcceptanceTestBase;
+import tech.pegasys.teku.test.acceptance.dsl.BesuNode;
+import tech.pegasys.teku.test.acceptance.dsl.TekuNode;
+import tech.pegasys.teku.test.acceptance.dsl.tools.deposits.ValidatorKeystores;
+
+public class OptimisticSyncPostMergeAcceptanceTest extends AcceptanceTestBase {
+  private static final String NETWORK_NAME = "less-swift";
+
+  private final SystemTimeProvider timeProvider = new SystemTimeProvider();
+  private BesuNode executionNode1;
+  private BesuNode executionNode2;
+  private TekuNode tekuNode1;
+  private TekuNode tekuNode2;
+
+  @BeforeEach
+  void setup() throws Exception {
+    final int genesisTime = timeProvider.getTimeInSeconds().plus(10).intValue();
+    executionNode1 =
+        createBesuNode(
+            config ->
+                config
+                    .withMiningEnabled(true)
+                    .withMergeSupport(true)
+                    .withP2pEnabled()
+                    .withGenesisFile("besu/preMergeGenesis.json"));
+    executionNode1.start();
+    executionNode2 =
+        createBesuNode(
+            config ->
+                config
+                    .withMergeSupport(true)
+                    .withP2pEnabled()
+                    .withGenesisFile("besu/preMergeGenesis.json")
+                    .withStaticPeers(executionNode1));
+    executionNode2.start();
+
+    final int totalValidators = 4;
+    final ValidatorKeystores validatorKeystores =
+        createTekuDepositSender(NETWORK_NAME)
+            .sendValidatorDeposits(executionNode1, totalValidators);
+    tekuNode1 =
+        createTekuNode(
+            config ->
+                configureTekuNode(config, executionNode1, genesisTime)
+                    .withValidatorProposerDefaultFeeRecipient(
+                        "0xFE3B557E8Fb62b89F4916B721be55cEb828dBd73")
+                    .withValidatorKeystores(validatorKeystores));
+    tekuNode1.start();
+    tekuNode2 =
+        createTekuNode(
+            config ->
+                configureTekuNode(config, executionNode2, genesisTime)
+                    .withInteropValidators(0, 0)
+                    .withPeers(tekuNode1));
+    tekuNode2.start();
+  }
+
+  @Test
+  void shouldSwitchToOptimisticSyncAfterMergeWhenExecutionEngineIsSyncing() throws Exception {
+    tekuNode2.waitForGenesis();
+
+    // Reset execution client's DB after the merge and leave it without any chance to sync
+    tekuNode2.waitForNonDefaultExecutionPayload();
+    executionNode2.removePeer(executionNode1);
+    executionNode2.waitForRestartWithEmptyDatabase();
+
+    tekuNode2.waitForOptimisticBlock();
```

Comment:
So I think this looks good now - maybe I was misunderstanding the flow of the previous test which is why my comment wasn't clear.

So from what I can see we have the EL (executionNode2) in sync so that tekuNode2 can transition through the merge block normally.  Then we reset executionNode2 and ensure it can't sync and teku should continue importing blocks optimistically.

It may be worth then re-adding the peer to executionNode2 and we can then wait for Teku's head block to not be optimistic anymore which would check that the EL is syncing properly.  If it makes the test take too long it may not be worth it, but if it can re-sync reasonably quickly it could be a useful test.

Reply:
Ok, I will add this, don't think it will take long, but maybe I'm wrong. 

My question is about another, `OptimisticSafeSlotsSyncPostMergeAcceptanceTest` test. I had following flow there:
Start:
Teku1 (with vals) -> Besu1 (mining)
Teku2 (no vals, connected to Teku1) -> Besu2 (no mining, connecting to Besu1)
After BeaconChain genesis:
Besu2 disconnects from Besu1 (it's pre-merge, and I agree with you that nothing guarantee that I'm in time until the merge there, plus I should clarify that it happens on Bellatrix, I guess), now it's syncing with no luck to sync
After that in `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` Teku2 should start to import blocks and pass the merge optimistically. Your comment was on that test. Plus there is an issue with it that Teku will not start import in `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` because `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` is more than 2, so we drop payload re-execution and have gap in the chain. We request non-gossip sync only on `StatusMessage` exchange which is once in 5 minutes. So it will start to import optimistically only in 5 minutes, not after 4 slots which is `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` if re-execution is dropped on the moment when safe slots are passed.

In `OptimisticSafeSlotsSyncPostMergeAcceptanceTest` I wanted to test second optimistic import condition which could let us even pass merge optimistically. It's not very good idea to wait 5 minutes in test, so I thought you had some idea how to rework it.

Reply:
So my understanding of the test is that we have a teku node which gets a SYNCING response from it's EL for the merge transition block (first one with a non-default payload) and it should pause imports until `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` then proceed to optimistically import.

Given there's no need to have an EL for blocks prior to the merge transition, my suggestion would be to simply start Besu2 with no peers and it will remain at genesis and return SYNCING all the time. Then we sync teku normally. If we later wanted besu to sync we could then add the peer but that's probably not necessarily.

`SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` is hard coded to 128 which is about 12 minutes on minimal, and 6.4 minutes on the `swift` config. So that will be the determining factor in how long it has to wait.  A test that takes that long isn't going to be valuable though so I'd suggest we just drop this test. Optimistically syncing through the merge is a more detailed case we'd be better off testing with integration tests.

Reply:
> Given there's no need to have an EL for blocks prior to the merge transition, my suggestion would be to simply start Besu2 with no peers and it will remain at genesis and return SYNCING all the time.

Teku will wait forever for BeaconChain genesis event in this case, it couldn't verify it

> SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY is hard coded to 128 

I've override it to 4 blocks and it works, the issue is in re-execution strategy. We stop re-execution after 2 slots, so we have chain:
SUCCESS <- SUCCESS <-[nothing]<- UNKNOW_PARENT <- UNKNOWN_PARENT
We are unable to pass it on gossip, we try to sync after `StatusMessage` exchange which is once in 5 minutes. So even with SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY == 4 we are importing blocks only after 5 minutes. I just want to clarify that we are ok with this behavior (as we have a lot of peers on mainnet and some `StatusMessage` arrives in seconds there and we need to protect EL from bad re-execution). If it's normal, I just drop safe slots test because it will take 6 minutes which is not acceptable

Reply:
> Teku will wait forever for BeaconChain genesis event in this case, it couldn't verify it

Ah, it probably makes sense to switch from sending deposits for the validators to just using interop mode. We don't need to be testing genesis generation here so can simplify things and speed up the test by using interop mode.

> I've overrided it to 4 slots, we have command-line parameter for it and it works, the issue is in re-execution strategy. We stop re-execution after 2 slots, after that we have a chain:
SUCCESS <- SUCCESS <-[nothing]<- UNKNOWN_PARENT <- UNKNOWN_PARENT
We are unable to pass it on gossip, we try to sync after StatusMessage exchange which is once in 5 minutes with every node, but we have 1 peer in test, so exactly 5 minutes. Even with SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY == 4 we are importing blocks only after 5 minutes. I just want to clarify that we are ok with this behavior (as we have a lot of peers on mainnet and some StatusMessage arrives in seconds there and we need to protect EL from bad re-execution). If it's normal, I just drop safe slots test because it will take 6 minutes which is not acceptable

This is definitely not behaviour we want, but continuing to try re-executing the block is not the right solution.  We may need to adjust our gossip handling to ensure it retries execution when we get a new descendent of a failed block (in a separate PR).